neuralpp/symbolic/expression.py

```diff
@@ -117,22 +122,11 @@ def syntactic_eq(self, other) -> bool:
             ), AtomicExpression(
                 base_type=other_base_type, atom=other_atom, type=_
             ):
-                # TODO: fix this
+                # TODO: fix this. If SymPyExpression typing system is fixed, `self_type == other_type` should work
                 # return self_base_type == other_base_type and self_type == other_type and self_atom == other_atom
                 return self_base_type == other_base_type and self_atom == other_atom
-            case (  # TODO: remove this case, which is taking care of semantic rather than syntactic equality
-                    # At the moment the treatment of Polynomials depends on this because they are represented
-                    # as applications of the identity function. Once we represent them in their own class,
-                    # we can remove this case.
-                     FunctionApplication(function=Constant(value=functions.identity), arguments=identity_arguments),
-                     the_other
-                 ) | \
-                 (
-                     the_other,
-                     FunctionApplication(function=Constant(value=functions.identity), arguments=identity_arguments)
-                 ):
-                assert len(identity_arguments) == 1
-                return identity_arguments[0].syntactic_eq(the_other)
+            case (SymPyPoly(poly=poly), the_other) | (the_other, SymPyPoly(poly=poly)):
```

Comment:
We still need a special case for Poly here, since `Poly` should be invisible. I believe `Poly(x+1)` should be defined as syntactically equal to `x+1`.

Reply:
I think that is a good solution for now, but I think we will need to redesign `syntactic_eq` to be more modular at some point. In principle, `Expression` should not know about all its subclasses; people should be able to define their own new subclasses of `Expression` without needing to modify `expression.py`. So `syntactic_eq` should be an abstract method and each class takes care of its own comparisons. One trouble is that some classes may not know about newer classes they can be compared against (for example `FunctionApplication` can be compared to `SymPyPoly` but it does not know about the latter). One way to do that would be for each implementation of `A.syntactic_eq(B)` to run `B.syntactic_eq(A)` if `B` is an unknown class. We can extend the return type of the method to `None` if it does not know how to compare. So when confronted with an unknown class, the method calls the inverse order and, if it still gets `None`, then it finally returns `None`. I think I will take a stab at implementing this.

Reply:
I think that is a good solution for now, but I think we will need to redesign `syntactic_eq` to be more modular at some point. In principle, `Expression` should not know about all its subclasses; people should be able to define their own new subclasses of `Expression` without needing to modify `expression.py`. So `syntactic_eq` should be an abstract method and each class takes care of its own comparisons. One trouble is that some classes may not know about newer classes they can be compared against (for example `FunctionApplication` can be compared to `SymPyPoly` but it does not know about the latter). One way to do that would be for each implementation of `A.syntactic_eq(B)` to run `B.syntactic_eq(A)` if `B` is an unknown class. We can extend the return type of the method to `None` if it does not know how to compare. So when confronted with an unknown class, the method calls the inverse order and, if it still gets `None`, then it finally returns `None`. I think I will take a stab at implementing this.